File: sdk-node/examples/forms/error-handling.ts
/**
 * Demonstrate the typed error tree on the forms surface. Each subclass
 * extends `ThreeCommonError`; branch with `instanceof`.
 *
 * Run:
 *   npx tsx examples/forms/error-handling.ts
 */

import {
  ThreeCommon,
  ThreeCommonAuthError,
  ThreeCommonNotFoundError,
  ThreeCommonRateLimitError,
  ThreeCommonValidationError,
} from '@3common/sdk'

const client = new ThreeCommon({
  apiKey: '3co_your_api_key_here',
})

try {
  // Expect: "validation: Name cannot be blank"
  await client.forms.create({ name: '', type: 'standalone' })
} catch (err) {
  if (err instanceof ThreeCommonValidationError) {
    console.warn(`validation: ${err.message}`)
  } else if (err instanceof ThreeCommonNotFoundError) {
    console.warn('form or element not found')
  } else if (err instanceof ThreeCommonAuthError) {
    console.warn('bad or expired API key')
  } else if (err instanceof ThreeCommonRateLimitError) {
    const wait = err.retryAfterSeconds ?? 30
    console.warn(`rate limited; retry in ${String(wait)}s`)
  } else {
    throw err
  }
}

File: sdk-python/examples/forms/error_handling.py
"""Run with: python examples/forms/error_handling.py"""

from __future__ import annotations

from threecommon import (
    AuthError,
    NotFoundError,
    RateLimitError,
    ThreeCommon,
    ValidationError,
)
from threecommon.forms import CreateBody


def main() -> None:
    with ThreeCommon(api_key="3co_your_api_key_here") as client:
        try:
            client.forms.create(CreateBody(name="Registration", type="standalone"))
        except ValidationError as e:
            print(f"validation: {e.message}")
        except NotFoundError:
            print("form or element not found")
        except AuthError as e:
            print(f"auth failed: check your API key (code={e.code})")
        except RateLimitError as e:
            wait = e.retry_after_seconds or 30
            print(f"rate limited; waiting {wait}s before retry")


if __name__ == "__main__":
    main()
